### ISSUE TITLE
autolog/logging should recreate/reopen log files each log event

### ISSUE THREAD

**Issue by midnightmagic** [irssi/irssi#541]
This would make things like autopath automatically catch up with new settings changes without cycling it; it would make atomic log rotation for long-lived files possible (via hardlink-then-delete) and would solve the problem of stale filehandles keeping open deleted files.

This is feasible because IRC lines are limited in terms of flow.

**Comment by ailin-nemui** [comment]
you can use the built in date based log rotation?

**Comment by ailin-nemui** [comment]
Closing for inactivity

**Comment by midnightmagic** [comment]
Sorry. I didn't realize you'd responded.

No, you can't. Irssi keeps open long-lived filehandles so external tools which truncate the logfile will not function with Irssi. There's no reason to keep the filehandles open for so long--if you open/append to the logfiles it would make irssi transparently more compatible with other tools (like logrotate for example) and backup tools which otherwise can't atomically pull copies of Irssi's logs.

It's a completely transparent change in behaviour to the end-user, except suddenly Irssi would no longer (basically ever) be writing to a deleted file, and lines wouldn't be lost.

:-)

**Comment by dequis** [comment]
Emphasis in "built in". See the last example in https://irssi.org/documentation/startup/#logging

> You can make irssi automatically rotate the logs by adding date/time formats to the file name. The formats are in “man strftime” format. For example
> 
>      /SET autolog_path ~/irclogs/%Y/$tag/$0.%m-%d.log

**Comment by midnightmagic** [comment]
Yeah, but the way the current logic works is that it internally rebuilds the filename each hour, and at that time it checks to see if the logfiles need rotation:

log.c:213 (log_write_rec())
        tm = localtime(&log->last);
        day -= tm->tm_mday; /* tm breaks in log_rotate_check() .. */
        if (tm->tm_hour != hour) {
                /* hour changed, check if we need to rotate log file */
                log_rotate_check(log);
        }   

Then in log_rotate_check():
        if (g_strcmp0(new_fname, log->real_fname) != 0) {
                /* rotate log */
                log_stop_logging(log); 
                signal_emit("log rotated", 1, log);
 
                log_start_logging(log);
        }    

I'm saying none of this logic is necessary, there doesn't need to be a timestamp check, there doesn't need to be a held-open set of file handles. All this can be constructed on-the-fly, and then the log appended to, and this would allow third-party atomic log rotation and logfile backups--right now if you externally copy away the logfiles, you get at least one file which is still being written-to, which means external software working with the logfiles can't do anything with the logfiles. It currently has to be smart enough to recognize when an old file it touched has been updated by Irssi, and as far as I can tell there's no way it can tell Irssi to reopen/re-seek its log files without a user script or a custom hook helping.

At the moment, I am in fact using that last example you're referring to in my Irssi installation and I have an rsync which copies the log files away to a semi-permanent storage area—but surely the solution is to do actual log rotation, and the best way to do that is to rip out all this log rotation logic and emulate it with per-record open-append log writing of the type I'm suggesting, which to the end-user would be nearly transparent.

The only drawback I can think of to my suggestion is that with the long-held writer locks you currently hold, it prevents interleaved Irssi logfile appends by two Irssi instances.

:-)

**Comment by ailin-nemui** [comment]
layman speaking, reopening the files sounds wasteful for the general case when noone rotated them away?

**Comment by midnightmagic** [comment]
It isn't; the volume of logging is minimal for IRC but at any rate on large commercial software which wrote tens of thousands of lines per second in this way on reasonable hardware, really the only issue is that on some filesystem types the seek to the end of the file was slow when the file exceeded the tens or hundreds of GB size threshold. Meanwhile the total filehandles held open decreases to zero for the usual case, or during lulls in conversation, and reduces to one maximum for short periods since logging is serial isn't it..?

**Comment by dequis** [comment]
I think we need some benchmarks here, particularly with high traffic and slow filesystems ("300 channels in raspberry pi sdcard" kind of stuff)

**Comment by cdhowie** [comment]
I'm running into this issue, too.  Perhaps we could investigate a compromise, where irssi will react to SIGUSR1 by closing all of its open log file handles?

**Comment by midnightmagic** [comment]
It's not a huge deal for me—it just means the logfile rotations I use are stale by at most 24 hours, and I can delete the old logfiles with about as much effort as it would take to rotate them and name them myself. A benchmark describing the speed with which lines can be appended to logfiles with this method would likely go a long way to convincing the Irssi devs this is safe to do.

**Comment by ailin-nemui** [comment]
having some way to make irssi reopen the logs sounds sensible, iirc some other apps also do this

**Comment by dterrahe** [comment]
This would also resolve my issue where if I need to restart irssi (due to a reboot) it will overwrite the current logfile (I use 1 file per day with autolog_path = "/mnt/Expansion/Staging/irclogs/$tag.$0.%Y-%m-%d.log"; )

**Comment by knofte** [comment]
+1 to making logrotation work.
When you have single files stretching hundreds of megs or even up in the gigs, it's really messy that you can't move them, tar them or such because "file changed while we read it".

Makes loghandling very difficult.

Let's please add the suggested change.
